### PR TITLE
Allow Dialogs to have a MainWindow independent config

### PR DESCRIPTION
**Pull request by mayman99**
Signed-off-by: Mohamad <mohamad@openrobotics.org>
🎉 New feature 

This PR allows Dialogs to read and write to the config file.
The motivation for this PR is, all the current reading/writing of the `Config` is implemented in `MainWindow`, and the Application has not been initialized yet then there is not `MainWindow` object.

a dialog config looks like `<dialog name=\"dialog_name\" attribute_1=\"true\"/>"`

Workflow:

1. Set dialog name `SetName`
2. set the dialog default config using `SetDefaultConfig`
3. Read an attribute of the dialog config using `ReadAttribute`
4. if the dialog config doesn't exist in the existing config file
    5. append the dialog default config to the config file
    6. read from the written config
7.  if the dialog config exists in the existing config file, read it and return the value as string
8. update an attribute value in the dialog config using `WriteAttributre`

The changes here are used in [this PR](https://github.com/gazebosim/gz-sim/pull/1536).
## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

**Review by chapulina — COMMENTED** [review]
> The motivation for this PR is, all the current reading/writing of the Config is implemented in MainWindow, and the Application has not been initialized yet then there is not MainWindow object.

I'm not sure I understand this part. The application has to be initialized before the dialog. Can't we use leave the config path handling to the `Application`? It would be nice if we didn't need another config file and could reuse the same one. The `<dialog>` elements could be siblings of the `<window>` element.

---

Have you seen the `WindowConfig` struct? That's used to hold an in-memory representation of an XML config file, load from a file and write to it. It would be interesting to follow the same pattern for consistency of the API, creating a new `DialogConfig` struct.

**Comment by mayman99** [comment]
Turning the Dialog class to be more like the MainWindow, like to have a similar API is better than coming up with a new API like how I just did on the PR.
I'll be modifying it to have a similar `DialogConfig` struct and a similar API, thanks for the suggestion

**Comment by mayman99** [comment]
Updated the PR per VC, except of using templates with `ReadAttributeValue`, as it results in a seg-fault when using in `gazebo`, which needs further debugging to figure out why.

File: src/Dialog.cc [hunk 1]
```diff
@@ -75,3 +78,153 @@ QQuickItem *Dialog::RootItem() const
...
+  // Write config file
+  tinyxml2::XMLPrinter printer;
+  doc.Print(&printer);
+
+  std::string config = printer.CStr();
+  std::ofstream out(_path.c_str(), std::ios::out);
+  if (!out)
+  {
+    std::string str = "Unable to open file: " + _path;
+    str += ".\nCheck file permissions.";
```
Comment (chapulina) [inline src/Dialog.cc:124]: The `str` variable isn't used anywhere, did you mean to print an error?

File: src/Dialog.cc [hunk 2]
```diff
@@ -75,3 +78,153 @@ QQuickItem *Dialog::RootItem() const
...
+      missingDoc.Print(&defaultPrinter);
+    }
+  }
+
+  // Write config file
+  tinyxml2::XMLPrinter printer;
+  doc.Print(&printer);
+
+  // Don't write the xml version decleration if file exists
+  if (configExists){
```
Comment (chapulina) [inline src/Dialog.cc:211]: Style nit

```suggestion
  if (configExists)
  {
```

File: src/Dialog.cc [hunk 3]
```diff
@@ -75,3 +78,153 @@ QQuickItem *Dialog::RootItem() const
...
+  igndbg << "Setting Quick start dialog default config." << std::endl;
+
+  config += printer.CStr();
+  config += defaultPrinter.CStr();
+  std::ofstream out(_path.c_str(), std::ios::out);
+  if (!out)
+  {
+    std::string str = "Unable to open file: " + _path;
+    str += ".\nCheck file permissions.";
+    igndbg << str << std::endl;
```
Comment (chapulina) [inline src/Dialog.cc:223]: This looks like an error

```suggestion
    ignerr << str << std::endl;
```

You could also save a copy and just stream the error directly to `ignerr` without going through an intermediate variable `str`

File: src/Dialog.cc [hunk 4]
```diff
@@ -75,3 +78,153 @@ QQuickItem *Dialog::RootItem() const
...
+    configExists = false;
+    doc.Parse(this->dataPtr->config.c_str());
+    // Process each dialog
+    for (auto dialogElem = doc.FirstChildElement("dialog");
+      dialogElem != nullptr;
+      dialogElem = dialogElem->NextSiblingElement("dialog"))
+    {
+      if(dialogElem->Attribute("name") == this->objectName().toStdString())
+      {
+        value = dialogElem->Attribute(_attribute.c_str());
```
Comment (chapulina) [inline src/Dialog.cc:162]: Currently, if there are multiple dialog elements with the same name, the function will return the last of them. I think that behaviour is ok. Can you document it?

File: src/Dialog.cc [hunk 5]
```diff
@@ -75,3 +78,153 @@ QQuickItem *Dialog::RootItem() const
...
+    {
+      return "";
+    }
+
+    // Process each existing dialog
+    for (auto dialogElem = doc.FirstChildElement("dialog");
+      dialogElem != nullptr;
+      dialogElem = dialogElem->NextSiblingElement("dialog"))
+    {
+      if(dialogElem->Attribute("name") == this->objectName().toStdString())
```
Comment (chapulina) [inline src/Dialog.cc:179]:
```suggestion
      if (dialogElem->Attribute("name") == this->objectName().toStdString())
```

File: src/Dialog_TEST.cc [hunk 6]
```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (C) 2017 Open Source Robotics Foundation
```
Comment (chapulina) [inline src/Dialog_TEST.cc:2]:
```suggestion
 * Copyright (C) 2022 Open Source Robotics Foundation
```

File: src/Dialog_TEST.cc [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      "allow");
+    EXPECT_EQ(allow, "");
+  }
+
+  // Read a non existing attribute
+  {
+    dialog->setObjectName("quick_menu");
+    dialog->SetDefaultConfig(std::string(
+      "<dialog name=\"quick_menu\" show=\"true\"/>"));
+    std::string allow = dialog->ReadAttributeValue(app.DefaultConfigPath(),
```
Comment (chapulina) [inline src/Dialog_TEST.cc:64]: You could check that the config file didn't exist, and then it was created by this function

File: src/Dialog.cc [hunk 8]
```diff
@@ -75,3 +78,153 @@ QQuickItem *Dialog::RootItem() const
...
+  else
+  {
+    auto success = !doc.LoadFile(_path.c_str());
+    if (!success)
+    {
+      return "";
+    }
+
+    // Process each existing dialog
+    for (auto dialogElem = doc.FirstChildElement("dialog");
```
[the remark this answers is not in the document]
Comment (mayman99) [reply src/Dialog.cc:172]: maybe I could use a lambda to hold the loop :thinking:

File: include/ignition/gui/Dialog.hh [hunk 9]
```diff
@@ -55,6 +56,28 @@ namespace ignition
...
+      /// \param[in] _path config path
+      /// \param[in] _attribute XMLElement attribute name
+      /// \param[in] _value XMLElement attribute value
+      /// \return true if written to config file
+      public: bool UpdateConfigAttribute(
+        const std::string &_path, const std::string &_attribute,
+        const bool _value) const;
+
+      /// \brief Gets a config attribute value, if not found in config
+      /// \brief write the default in the config and get it.
```
Comment (chapulina) [inline include/ignition/gui/Dialog.hh:73]:
```suggestion
      /// write the default in the config and get it.
```

This is a continuation of the brief above. Same for the next line.

File: src/Dialog.cc [hunk 10]
```diff
@@ -75,3 +80,141 @@ QQuickItem *Dialog::RootItem() const
...
+  std::string value {""};
+  std::string config = "<?xml version=\"1.0\"?>\n\n";
+  tinyxml2::XMLPrinter defaultPrinter;
+  bool configExists{true};
+  std::string dialogName = this->objectName().toStdString();
+
+  auto Value = [&_attribute, &doc, &dialogName]()
+  {
+    // Process each dialog
+    // If multiple attributes share the same name, return the last one
```
Comment (chapulina) [inline src/Dialog.cc:151]: It looks like the new logic actually returns the first one, right?

File: src/Dialog_TEST.cc (hunk 7 above)
Comment (chapulina) [reply src/Dialog_TEST.cc]: Thanks for adding it for the first test case. Can you add those checks for all the others? It's very useful to be explicit about the exact case that's being tested.

Comment (mayman99) [reply src/Dialog_TEST.cc]: Make sense :+1:

File: src/Dialog.cc (hunk 10 above)
Comment (mayman99) [reply src/Dialog.cc]: copy pasting comments without re-reading :upside_down_face: 
thanks for all the catches!

**Comment by mayman99** [comment]
@chapulina green :+1: 
I didn't have time to convert the Read/Update functions to use template functions (I did but it results in a seg fault when used in `gz-sim`, my maintenance time was not enough to find the problem)

**Review by chapulina — APPROVED** [review]

**Pull request merged**